### ISSUE TITLE
NetCore: DataValueEditor.View no longer resolves virtual path

### ISSUE THREAD

**Issue by leekelleher** [umbraco/Umbraco-CMS#10265]
When developing a custom property-editor using the `DataEditorAttribute` for v9.0.0-beta001, if the `View` path is virtual, (e.g. prefixed with `~/`) it is not resolved.

In my Contentment package, many of the property-editors use the `DataEditorAttribute`, an example here: <https://github.com/leekelleher/umbraco-contentment/blob/1.4.2/src/Umbraco.Community.Contentment/DataEditors/CodeEditor/CodeEditorDataEditor.cs#L16>

> Please ignore that I'm using string constants, the value would be `"~/App_Plugins/Contentment/editors/code-editor.html"`.
> For an example from the Umbraco documentation, see here: <https://our.umbraco.com/documentation/Tutorials/Creating-a-Property-Editor/#setting-up-a-property-editor-with-c>

When I used this editor on a DocumentType, I get the following error in the back-office...

> **Request error: The URL returned a 404 (not found):**
> **`~/App_Plugins/Contentment/editors/code-editor.html`**

In v8, the `View` path is resolved using the `IOHelper.ResolveVirtualUrl` helper method.

<https://github.com/umbraco/Umbraco-CMS/blob/v8/contrib/src/Umbraco.Core/PropertyEditors/DataValueEditor.cs#L78>

For v9.0.0-beta001, that has been removed.

<https://github.com/umbraco/Umbraco-CMS/blob/release-9.0.0-beta001/src/Umbraco.Core/PropertyEditors/DataValueEditor.cs#L92>

---

I'm aware in most cases that I should be injecting the `IIOHelper` dependency into my property-editor, but I'm unsure how that would work with the `DataEditorAttribute`?

Would we considered this to be a regression bug?

Reproduction
------------

If you're filing a bug, please describe how to reproduce it. Include as much
relevant information as possible, such as:

### Bug summary

_(At this level, I kinda feel that I'm teaching you to suck eggs here?)_ 🤷 

- Create a custom property-editor using the C# approach: see <https://our.umbraco.com/documentation/Tutorials/Creating-a-Property-Editor/#setting-up-a-property-editor-with-c>
- For the `View` property, use a path that starts with `"~/"`.
- Compile/build/whatever.
- Create the DataType.
- Add new property to DocumentType - **At this point, the error will occur.**
- Create/open a content node (of the above DocumentType) - **you'll get the error here too.**

### Specifics

- Umbraco version: v9.0.0-beta001
- Browser: Chrome vLatest
- I haven't checked on any other browser.
- Error message outlined above, e.g. "Request error: The URL returned a 404 (not found)".

### Steps to reproduce

_(Outlined above)_

### Expected result

I'd expect the relative/virtual `View` path to have been resolved, as it did in v8.

### Actual result

_(Outlined above)_

**Comment by bergmania** [comment]
Hi @leekelleher.. Thanks for reporting..

The functionality is removed from the property setter on purpose. But I think it is important your use case still works. I think this should be fairly easy to reintroduce so the virtual path is resolved when using the Attribute.

**Comment by jemayn** [comment]
@bergmania 

Are we sure this is fixed? I am getting the same issue on 9.1.2, I have an editor like this:

```
using Skybrud.Integrations.BorgerDk;
using Umbraco.Cms.Core.IO;
using Umbraco.Cms.Core.PropertyEditors;

namespace Skybrud.Umbraco.BorgerDk.PropertyEditors {

    [DataEditor("Skybrud.BorgerDk", EditorType.PropertyValue, "Skybrud Borger.dk", "~/App_Plugins/Skybrud.BorgerDk/Views/Editor.html", ValueType = ValueTypes.Json, Group = "Skybrud.dk", Icon = "icon-school")]
    public class BorgerDkPropertyEditor : DataEditor {
        private readonly IIOHelper _iOHelper;

        public BorgerDkPropertyEditor(IDataValueEditorFactory dataValueEditorFactory, IIOHelper iOHelper) : base(dataValueEditorFactory)
        {
            _iOHelper = iOHelper;
        }

        /// <inheritdoc />
        protected override IConfigurationEditor CreateConfigurationEditor() => new BorgerDkConfigurationEditor(_iOHelper);

    }

    public class BorgerDkConfigurationEditor : ConfigurationEditor<BorgerDkConfiguration> {

        public BorgerDkConfigurationEditor(IIOHelper iOHelper) : base(iOHelper) { }

    }

    public class BorgerDkConfiguration {

        [ConfigurationField("municipality", "Municipality", "~/App_Plugins/Skybrud.BorgerDk/Views/Municipality.html", Description = "Select the municipality to be used.")]
        public BorgerDkMunicipality Municipality { get; set; }

        [ConfigurationField("allowedTypes", "Allowed types", "~/App_Plugins/Skybrud.BorgerDk/Views/AllowedTypes.html", Description = "Select the element types that should be allowed.")]
        public string[] AllowedTypes { get; set; }

    }

}
```

And both the view from the `[DataEditor]` and the `[ConfigurationField]` attributes are giving 404s in the backoffice, even though the files exist on those paths in the content root.

**Comment by jemayn** [comment]
Nvm, found the issue - this seems to work as expected 🙂

For anyone else running into similar problems in the future - NuGet packages in v9 need a targets file that copies over the App_Plugins files - even if they are there in the solution they are apparently not in the right place. 

In my instance I added this in my csproj:
```
<ItemGroup>
    <Content Include="../build/Skybrud.Umbraco.BorgerDk.targets" PackagePath="build\Skybrud.Umbraco.BorgerDk.targets" />
```

And this targets file:
```
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">

  <!--Based on: https://github.com/bergmania/Bergmania.OpenStreetMap/blob/main/Bergmania.OpenStreetMap/build/Bergmania.OpenStreetMap.targets -->

  <PropertyGroup>
    <SkybrudBorgerDkContentFilesPath>$(MSBuildThisFileDirectory)..\content\App_Plugins\Skybrud.BorgerDk\**\*.*</SkybrudBorgerDkContentFilesPath>
  </PropertyGroup>

  <Target Name="CopySkybrudBorgerDkAssets" BeforeTargets="Build">
    <ItemGroup>
      <SkybrudBorgerDkContentFiles Include="$(SkybrudBorgerDkContentFilesPath)" />
    </ItemGroup>
    <Message Text="Copying Skybrud.Umbraco.BorgerDk files: $(SkybrudBorgerDkContentFilesPath) - #@(SkybrudBorgerDkContentFiles->Count()) files"  Importance="high" />
    <Copy
      SourceFiles="@(SkybrudBorgerDkContentFiles)"
      DestinationFiles="@(SkybrudBorgerDkContentFiles->'$(MSBuildProjectDirectory)\App_Plugins\Skybrud.BorgerDk\%(RecursiveDir)%(Filename)%(Extension)')"
      SkipUnchangedFiles="true" />
  </Target>

  <Target Name="ClearSkybrudBorgerDkAssets" BeforeTargets="Clean">
    <ItemGroup>
      <SkybrudBorgerDkDir Include="$(MSBuildProjectDirectory)\App_Plugins\Skybrud.BorgerDk\" />
    </ItemGroup>
    <Message Text="Clear old Skybrud.Umbraco.BorgerDk files"  Importance="high" />
    <RemoveDir Directories="@(SkybrudBorgerDkDir)"  />
  </Target>

</Project>
```
